src/hotspot/share/classfile/classFileParser.cpp

```diff
@@ -4582,6 +4586,14 @@ void ClassFileParser::check_super_class_access(const InstanceKlass* this_klass,
       return;
     }
 
+    // A value class's super must be either java.lang.Object or have ACC_PERMITS_VALUE set.
+    if (this_klass->access_flags().is_value_class() &&
+        super_ik->name() != vmSymbols::java_lang_Object() &&
+        !super_ik->is_permits_value_class()) {
+      classfile_icce_error("value class %s cannot inherit from class %s", super_ik, THREAD);
+      return;
+    }
+
```

Comment:
This is not the behavior written in the JVMS draft for value classes, which requires the direct super class of a value class to have the ACC_PERMITS _VALUE flag set. However, because we haven't changed java.lang.Object into an abstract class, java.lang.Object cannot have the ACC_PERMITS _VALUE flag set. Currently, I see no other solution than the one you have implemented, adding a special case when the direct super class is java.lang.Object. But it would be useful to add a comment in the source code and the CR to remember and track this discrepancy.

Reply:
Thanks for suggesting this.  I added a comment.